src/hypothesis/patterns.py

```diff
@@ -0,0 +1,549 @@
+            remaining_max_size = _sum_max_size(patterns[n:])
+        elif remaining_max_size is not None:
+            remaining_max_size = max(
+                0,
+                remaining_max_size - pattern._max_size,
+            )
+        if remaining_max_size is None:
+            pattern_min_size = 0
+        else:
+            pattern_min_size = max(
+                0,
+                min_size - result_len - remaining_max_size,
+            )
+        # Get the strategy and draw the result.
+        strategy = pattern._get_strategy(
+            pattern_min_size,
+            pattern_max_size,
+        )
+        result_part = draw(strategy)
+        result_parts.append(result_part)
+        result_len += len(result_part)
+    # All done!
+    return u''.join(result_parts)
+
+
+class ConcatPattern(Pattern):
+
+    def __init__(self, patterns):
+        super(ConcatPattern, self).__init__(
+            _sum_min_size(patterns),
+            _sum_max_size(patterns),
+        )
+        self._patterns = patterns
+
+    def __add__(self, other):
+        if isinstance(other, ConcatPattern):
+            # Flatten branches.
+            return ConcatPattern(self._patterns + other._patterns)
+        return super(ConcatPattern, self).__add__(other)
+
+    def _get_strategy_impl(self, min_size, max_size):
+        return _concat_strategy(self._patterns, min_size, max_size)
+
+
+class OneOfPattern(Pattern):
+
+    def __init__(self, patterns):
+        super(OneOfPattern, self).__init__(
+            min(pattern._min_size for pattern in patterns),
+            None if any(
+                pattern
+                for pattern
+                in patterns
+                if pattern._max_size is None
+            ) else max(
+                pattern._max_size
+                for pattern
+                in patterns
+            ),
+        )
+        self._patterns = patterns
+
+    def __or__(self, other):
+        if isinstance(other, OneOfPattern):
+            # Flatten branches.
+            return OneOfPattern(self._patterns + other._patterns)
+        return super(OneOfPattern, self).__or__(other)
+
+    def _get_strategy_impl(self, min_size, max_size):
+        return st.one_of([
+            pattern._get_strategy(min_size, max_size)
+            for pattern
+            in self._patterns
+            if ((max_size is None or pattern._min_size <= max_size) and
+                (pattern._max_size is None or pattern._max_size >= min_size))
+        ])
+
+
+def one_of(*patterns):
+    return reduce(or_, patterns)
+
+
+def maybe(pattern):
+    return _expect_pattern(pattern) | just(u'')
+
+
+class JustPattern(Pattern):
+
+    def __init__(self, value):
+        super(JustPattern, self).__init__(len(value), len(value))
+        self._value = value
+        self._strategy = st.just(self._value)
+
+    def _get_strategy_impl(self, min_size, max_size):
+        return self._strategy
+
+
+def just(value):
```

Comment:
This is a name conflict with the `just` strategy. What do you think about calling this `literal`?